integration_tests/src/main/python/prune_partition_column_test.py

```diff
@@ -127,3 +128,78 @@ def test_prune_partition_column_when_filter_fallback_project(spark_tmp_path, pru
                                                              filter_col, file_format):
     do_prune_partition_column_when_filter_project(spark_tmp_path, prune_part_enabled, file_format,
                                                   filter_col, gpu_project_enabled=False)
+
+def create_contacts_table_and_read(is_partitioned, format, data_path, expected_schemata, func, conf):
+    full_name_type = StructGen([('first', StringGen()), ('middle', StringGen()), ('last', StringGen())])
+    name_type = StructGen([('first', StringGen()), ('last', StringGen())])
+    contacts_data_gen = StructGen([
+        ('id', IntegerGen()),
+        ('name', full_name_type),
+        ('address', StringGen()),
+        ('friends', ArrayGen(full_name_type, max_length=10, nullable=False))], nullable=False)
+
+    brief_contacts_data_gen = StructGen([
+        ('id', IntegerGen()),
+        ('name', name_type),
+        ('address', StringGen())], nullable=False)
+
+    def contact_gen_df(spark, data_gen, partition):
+        gen = gen_df(spark, data_gen)
+        if is_partitioned:
+            return gen.withColumn('p', f.lit(partition))
```

Comment:
I am really confused by this part here. You are adding in a column to the data, but further down you also add in a column based off the directory structure as an actual partition. You are effectively adding in p twice. Is this intended? From the test they are doing the same thing, but it is really confusing. Please document it, and help us understand what is happening here because it is really odd.

Reply:
I'm guessing this is inspired by the setup in `SchemaPruningSuite`. I haven't how this is done in `SchemaPruningSuite`, but this does look like it can be streamlined.

Reply:
Ok, it took me a little while to understand what `SchcemaPruningSuite`'s two dataframes were doing.

I think the test setup's goal is to produce a single large table with two partitions, with schema differences between the partition. The test then checks that reads across the partitions work correctly, as does the column pruning.

The way they codify the schema differences is by writing a different dataframe into each partition. Here are the dataframes at play:
1. `Contacts` (with `FullName`, `pets`, `friends`, etc.).
2. `BriefContacts` (containing the first 3 columns from `Contacts`).
3. `ContactsWithDataPartition` (containing all columns from `Contact`, and a partition column `p`.)
4. `BriefContactsWithDataPartition` (i.e. all column from `BriefContact` and a partition column `p`.)

1 and 2 above are written to partitions `p=1` and `p=2` of the same table.
3 and 4 seem to be written to partitions `p=1` and `p=2` of a different table.

`SchemaPruningSuite` seems to be clumsy, in adding the partition column's value _in the data file_. This seems wrong.

I don't think this mistake needs repeating in our tests. We should probably leave out the `p` column from the data file write.

Reply:
Note: I would have set up this test quite differently:

```python
# First, create the table.
sql("CREATE TABLE my_test_table(id INT, name STRUCT<`first`:STRING...>, address STRING) PARTITIONED BY (p INT) STORED AS ORCFILE LOCATION '...'")

# Create temp input from `brief_contacts`. This will be our first partition.
brief_contacts_data_gen_df.createOrReplaceTempView("input")

# Insert `brief_contacts` into `p=1`.
sql("INSERT OVERWRITE TABLE my_test_table PARTITION (p='1') SELECT * FROM input")

# Add some columns to my_test_table, to match `contacts_df`.
sql("ALTER TABLE my_test_table ADD COLUMNS(pets INT, friends ARRAY<STRUCT<...>>, ...)")

# Create temp input from `contacts`, with all the "new" columns.
contacts_data_gen_df.createOrReplaceTempView("input")

# Insert `contacts` into `p=2`
sql("INSERT OVERWRITE TABLE my_test_table PARTITION (p='2') SELECT * FROM input")
```

This uses static partitioning. None of the data files under `my_test_table` will have the partition key persisted in them.

Reply:
@mythrocks and @revans2 thank you for providing feedback. I have gone back and forth between the two approaches on whether to include `p` as a part of the table or not. I settled for keeping it like Spark because 1. I wanted to keep it consistent, 2. I wasn't sure if I would miss some Spark magic by dropping `p` 

I will post my findings about what happens if we have `p` in the table and a partition with different values.

Reply:
> Might be nice to have a follow on issue to also look at what happens if we have a column in a file and as a partition, but they values don't match. Just curious if we do the same thing as spark does in those cases.

I tried this, the partition value overwrites the value in the dataframe. 

```
scala> val df = spark.read.parquet("/tmp/pyspark_tests/raza-linux-1-main-1813088-766074268/DATA/contacts/p=1")
df: org.apache.spark.sql.DataFrame = [id: int, name: struct<first: string, middle: string ... 1 more field> ... 3 more fields]

scala> df.selectExpr("*").where("p=4").count
res4: Long = 2048

scala> df.selectExpr("*").where("p=3").count
res5: Long = 0

scala> df.selectExpr("*").where("p=2").count
res6: Long = 0

scala> df.selectExpr("*").where("p=1").count
res7: Long = 0

scala> val df = spark.read.parquet("/tmp/pyspark_tests/raza-linux-1-main-1813088-766074268/DATA/contacts/")
23/07/26 19:41:25 WARN DataSource: Found duplicate column(s) in the data schema and the partition schema: `p`
df: org.apache.spark.sql.DataFrame = [id: int, name: struct<first: string, middle: string ... 1 more field> ... 3 more fields]

scala> df.selectExpr("*").where("p=2").count
res0: Long = 2048                                                               

scala> df.selectExpr("*").where("p=1").count
res1: Long = 2048

scala> df.selectExpr("*").where("p=3").count
res2: Long = 0

scala> df.selectExpr("*").where("p=4").count
res3: Long = 0

```

Reply:
> what happens if we have a column in a file and as a partition, but they values don't match. 

From the looks of it, in Apache Spark, the partition value is honoured.

I tried creating ORC files with different values of column `p`, and copied them into a new table partitioned on column `p`, into partition `p=1`.
On Apache Spark, with ORC, all the other columns (except `p`) are read correctly. The `p` column seems to be ignored, and the partition value is honoured.
This happens regardless of the position of `p` in the file.